expr/func_aspercent.go

```diff
@@ -0,0 +1,285 @@
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+					}
+				}
+				outSeries = append(outSeries, serie1)
+				cache[Req{}] = append(cache[Req{}], serie1)
+			}
+		}
+		return outSeries, nil
+	}
+
+	var totalsSerie models.Series
+	if math.IsNaN(s.totalFloat) && totals == nil {
+		totalsSerie = sumSeries(series)
+		if len(series) == 1 {
+			totalsSerie.Target = fmt.Sprintf("sumSeries(%s)", totalsSerie.Target)
+			totalsSerie.QueryPatt = fmt.Sprintf("sumSeries(%s)", totalsSerie.QueryPatt)
+			totalsSerie.Tags = map[string]string{"name": totalsSerie.Target}
+		}
+	} else if totals != nil {
+		if len(totals) == 1 {
+			totalsSerie = totals[0]
+		} else if len(totals) == len(series) {
+			sort.Sort(sortByTarget(series))
+			sort.Sort(sortByTarget(totals))
+			for i, serie1 := range series {
+				serie2 := totals[i]
+				serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+				serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+				serie1.Tags = map[string]string{"name": serie1.Target}
+				for i := range serie1.Datapoints {
+					serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
+				}
+				outSeries = append(outSeries, serie1)
+				cache[Req{}] = append(cache[Req{}], serie1)
+			}
+			return outSeries, nil
+		} else {
+			return nil, errors.New("asPercent second argument must be missing, a single digit, reference exactly 1 series or reference the same number of series as the first argument")
+		}
+	} else {
+		totalsSerie.QueryPatt = fmt.Sprint(s.totalFloat)
+	}
+
+	for _, serie := range series {
+		serie.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie.QueryPatt, totalsSerie.QueryPatt)
+		serie.Target = fmt.Sprintf("asPercent(%s,%s)", serie.Target, totalsSerie.QueryPatt)
+		serie.Tags = map[string]string{"name": serie.Target}
+		for i := range serie.Datapoints {
+			var totalVal float64
+			if len(totalsSerie.Datapoints) > i {
+				totalVal = totalsSerie.Datapoints[i].Val
+			} else {
+				totalVal = s.totalFloat
+			}
+			serie.Datapoints[i].Val = computeAsPercent(serie.Datapoints[i].Val, totalVal)
+		}
+		outSeries = append(outSeries, serie)
+		cache[Req{}] = append(cache[Req{}], serie)
+	}
+
+	return outSeries, nil
+}
+
+func computeAsPercent(in, total float64) float64 {
+	if math.IsNaN(in) || math.IsNaN(total) {
+		return math.NaN()
+	}
+	if total == 0 {
+		return math.NaN()
+	}
+	return in / total * 100
+}
+
+func groupSeriesByKey(series []models.Series, nodes []expr, keys *[]string) map[string][]models.Series {
+	keyedSeries := make(map[string][]models.Series)
+Loop:
+	for _, serie := range series {
+		key := aggKey(serie, nodes)
+		if _, ok := keyedSeries[key]; !ok {
+			keyedSeries[key] = []models.Series{serie}
+			for _, k := range *keys {
```

Comment:
I really don't see many cases where len(keys) > 100, or 1000 at most. This is done in Graphite now, so I'm not sure that a map is really necessary, it would only decrease readability of code.

Reply:
> it would only decrease readability of code.

Would it? The key building would go from:
```
for _, k := range *keys {
    if k == key {
        continue Loop
    }
}
*keys = append(*keys, key)
```

to:

```
(*keys)[key] = struct{}
```

You would also get to remove a loop label (which I find difficult to read)

Reply:
fair enough, changed!